### PR TITLE
Gradients of correlations

### PR DESCRIPTION
**Pull request by seshasaibehara**
- Added fadbad includes in `casm/clexulator/external` to be able to compile clexulator capable of doing derivatives
- ported over `grad_correlations` function from 1.X to c++ side and exposed it in python
- Tests are included for displacement degrees of freedom where the constant term polynomials are generated on square lattice using casm1.X
- The gradient values are compared against the analytical derivatives of the polynomials at a given displacement degrees of freedom

### CODE REVIEW

File: src/casm/clexulator/Correlations.cc [hunk 1]
```diff
@@ -613,5 +615,89 @@ std::vector<unsigned int> all_correlation_indices(Index n) {
...
+  // this const_cast is not great...
+  // but it seems like the only place passing const Clexulator is a problem and
+  // it is not actually changing clexulator before/after this function
+
+  const_cast<Clexulator &>(*m_clexulator)
+      .param_pack()
+      .set_eval_mode(corr_key, "DIFF");
+  const_cast<Clexulator &>(*m_clexulator)
+      .param_pack()
+      .set_eval_mode(dof_key, "DIFF");
```
Comment (seshasaibehara) [inline src/casm/clexulator/Correlations.cc:647]: @bpuchala any thoughts here?

Comment (bpuchala) [reply src/casm/clexulator/Correlations.cc:647]: I have no problem with it. There is at least one other place where we have to do a const_cast when evaluating the change in correlations, and as long as the const object is not changed at the end of the function it makes sense.

File: python/src/clexulator.cpp [hunk 2]
```diff
@@ -1161,7 +1161,22 @@ PYBIND11_MODULE(_clexulator, m) {
               Change in extensive correlations, relative to `reference_extensive_correlations`.
           )pbdoc",
            py::arg("key"), py::arg("new_value"),
-           py::arg("`reference_extensive_correlations"));
+           py::arg("`reference_extensive_correlations"))
```
Comment (bpuchala) [inline python/src/clexulator.cpp:1164]: Looks like a stray tick "`"

File: python/src/clexulator.cpp [hunk 3]
```diff
@@ -1161,7 +1161,22 @@ PYBIND11_MODULE(_clexulator, m) {
...
+          Calculate and return gradients of correlations
+
+          Parameters
+          ----------
+          key: str
+              Specifies the type of DoF
+
+          Returns
+          -------
+          gradients_of_correlations: np.ndarray
```
Comment (bpuchala) [inline python/src/clexulator.cpp:1176]: Can you please improve this docstring? 

For the `key` parameter, let's use `dof_key` and copy the text from one of the other methods.

Definitely need to describe the `gradients_of_correlations` array more: at least dimensions, dtype, and row/column meaning. For the type, I prefer: `numpy.ndarray[np.float64[m,n]]`, where `m` and `n` (or a more descriptive variable name) are described in the text.

File: python/tests/clexulator/test_diff_clex.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        \\Phi_{2} = dz_{0}^{2}
+        \\Phi_{3} = \\sqrt{1/2}(dx_{0}^{4} +dy_{0}^{4} )
+        \\Phi_{4} = \\sqrt{6}dx_{0}^{2} dy_{0}^{2}
+        \\Phi_{5} = \\sqrt{3}(dx_{0}^{2} dz_{0}^{2} +dy_{0}^{2} dz_{0}^{2} )
+        \\Phi_{6} = dz_{0}^{4}
+
+    Parameters
+    ----------
+    diff_clexulator : libcasm.clexulator.Clexulator
+    session_shared_datadir : str
```
Comment (bpuchala) [inline python/tests/clexulator/test_diff_clex.py:64]: Guess this is no longer needed.

File: python/src/clexulator.cpp (hunk 2 above)
Comment (bpuchala) [inline python/src/clexulator.cpp]: Looks like a stray tick "`"

File: src/casm/clexulator/Correlations.cc [hunk 5]
```diff
@@ -613,5 +615,89 @@ std::vector<unsigned int> all_correlation_indices(Index n) {
   return corr_indices;
 }
 
+/// Gradients of Correlations
+
+Eigen::MatrixXd const Correlations::grad_correlations(DoFKey const &key) {
```
Comment (bpuchala) [inline src/casm/clexulator/Correlations.cc:620]: Can you please add docs here too? Especially the output interpretation again.

**Review by bpuchala — CHANGES_REQUESTED** [review]
Looks good! I have a couple requests for documentation before we merge. Thanks!

**Review by bpuchala — APPROVED** [review]

**Pull request merged**